### ISSUE TITLE
Sound skips in JS

### ISSUE THREAD

**Issue by FireSlash** [HeapsIO/heaps#280]
**Haxe version:** 3.4.2
**Heaps version:** Git 7dece13 (Latest as of the creation of this issue)

Tested in Chrome and Edge. In both, the music skips forward at regular, short (~5 sec) intervals.

Repro case:
- Place an mp3 file named testmp3.mp3 in res/music/

```
class Main extends hxd.App {
    
    override function init() {

        var res = if( hxd.res.Sound.supportedFormat(Mp3) || hxd.res.Sound.supportedFormat(OggVorbis) ) hxd.Res.music.testmp3 else null;

        if( res != null ) 
        {
			trace("Playing "+res);
			var music = res.play(true);
		}

    }
    
    override function update(dt:Float) {
        
    }

    override function render(e:h3d.Engine) {
		
	}
    
    static function main() {
        hxd.Res.initEmbed();
        
        new Main();
    }
}

```

**Comment by FireSlash** [comment]
Gremlins may be afoot. 

For reasons I have yet to pin down, the behavior has changed: Now there's one much longer skip at the beginning of the song, at which point it plays the remainder without incident. If looping, the second loop plays fine.

Confirmed behavior with both Chrome and Edge again. Neither heaps or haxe have been updated, and I tried removing the two libs I installed since the last test to see if that made a difference.

As for the initial skip, I've been unable to pin down a clear cause, other than the exact skip amount appears to be consistent across multiple machines. I also noticed that if I load more content (such as some textures) after I call play(), I can shrink the resulting gap. 

I also verified this on a few other machines this time, and the behavior appears to be consistent. At least until the gremlins return.

**Comment by ncannasse** [comment]
Might be related to the way the sound needs to be loaded/streamed, could
you add some traces in hxd.snd functions to see where the delay might come
from?

**Comment by FireSlash** [comment]
Ok, I think I've nailed down the skipping bug.

getSourcei will return AL.STOPPED after about 5 seconds, which is the duration of our stream buffer coincidentally.

After further digging, I noticed that updateDuration is being called, which is good, but only after a few hours of head scratching did I realize that we're actually just slamming duration to 0 every time we call it, meaning after those 5 seconds are up getSourcei will assume we've played the whole file and return AL.STOPPED. This appears to confuse ALEmulator, which throws away all of it's buffers then just re-creates them (at a playedTime slightly further in the future), causing our skip.

Unfortunately the fix isn't quite as simple as commenting that line out, as updateDuration appears to produce a duration 3 times as long as the actual audio source. Oddly enough, this doesn't appear to have any kind of negative impact on the looping behavior, so it still "works fine". 

Still looking at the initial delay, but it's getting late and I need a fresh set of eyes.

**Comment by FireSlash** [comment]
I have a better handle on the initial skip now as well.

At a high level, the problem is that the sound system doesn't know whether the source is ready to supply it bytes or not. The event flow appears to be this:
- Sound::play() is called
- A new Mp3Data gets created, and begins to decode data via the JS native decodeAudioData async function
- ALEmulator asks for bytes to fill its buffer
- Mp3Data hasn't populated it's internal buffer yet, so it fills ALEmulator's target buffer with blank data
- This happens for two to three buffers worth of data before processBuffer gets called via decodeAudioData's callback
- By the third buffer or so, we start filling with real data; but three buffers in; which produces our skip.

My current workaround is to call sound::getData() in advance, which triggers processing ahead of time, though this is mostly just a hack.

**Comment by ncannasse** [comment]
Ok we might then want to delay the actual openAL driver calls until the
sound data is considered "ready"

**Comment by FireSlash** [comment]
I agree. With that change, It's also probably worth converting the buffer format in advance. Mp3Data stores its data as i32, but the HTML Audio API expects floats.

Right now, the buffer chunk we want gets converted on demand. This is unfortunately not a fast operation, taking about 3ms per channel per buffer on a high end machine (The whole copy operation spans about 9ms, so roughly 2/3 of it is spent converting the data).

It's not hard to imagine a scenario in which we would drop a frame here.

**Comment by thejustinwalsh** [comment]
**Haxe Version**: 4.0.0 (git build development @ 3018ab1)
**Heaps Version**: 0d7b4efed7d4141312a06ae117547a2784cf2fd2
**Browsers**: Safari 11.1, Chrome 66, and Firefox 59

I do not hear any audio in the Sound sample or in my own project. Could very well be a new issue, but seems related, as I know it is at least creating the context.

**Comment by ncannasse** [comment]
How, wonderful change in Chrome policy, didn't hear about it (pun intended)
https://developers.google.com/web/updates/2017/09/autoplay-policy-changes#webaudio

**Comment by ncannasse** [comment]
This should have been fixed with #437
